src/julia-syntax.scm

```diff
@@ -4015,30 +4039,40 @@ f(x) = yt(x)
                                        (table)
                                        (table)
                                        (null? (cadr e)) ;; only toplevel thunks have 0 args
-                                       interp opaq)))
+                                       interp opaq globals)))
              `(lambda ,(cadr e)
                 (,(clear-capture-bits (car (lam:vinfo e)))
                  () ,@(cddr (lam:vinfo e)))
                 (block ,@body))))
           ;; remaining `::` expressions are type assertions
           ((|::|)
-           (cl-convert `(call (core typeassert) ,@(cdr e)) fname lam namemap defined toplevel interp opaq))
+           (cl-convert `(call (core typeassert) ,@(cdr e)) fname lam namemap defined toplevel interp opaq globals))
           ;; remaining `decl` expressions are only type assertions if the
           ;; argument is global or a non-symbol.
           ((decl)
            (cond ((and (symbol? (cadr e))
                        (local-in? (cadr e) lam))
                   '(null))
                  (else
-                  (if (or (symbol? (cadr e)) (and (pair? (cadr e)) (eq? (caadr e) 'outerref)))
-                      (error "type declarations on global variables are not yet supported"))
-                  (cl-convert `(call (core typeassert) ,@(cdr e)) fname lam namemap defined toplevel interp opaq))))
+                  (cl-convert
+                    (let ((ref (binding-to-globalref (cadr e))))
+                      (and ref (get globals ref #f)
+                           (error (string "multiple type annotations for global \""
```

Comment:
I think we can remove this error and just let it get caught by the run time check. There are a few reasons for this. (1) Global declarations work like top-level effects, e.g.
```
let
    global const x = 1
    if false
        global const x = 2
    end
end
```
does not error since the second declaration doesn't execute.
(2) It disallows things that are allowed at the top level, e.g.
```
let
    global y::Int = 1
    global y::Int = 2
end
```
errors, but without the `let` it's fine.

Reply:
Though come to think of it, we should probably disallow type declarations on globals inside functions, like we do for `const` declarations.

Reply:
My idea was to mostly follow the behavior for type annotations in local scopes here, so that adding a `global x` inside a function should change as little as possible. I am not sure it's exactly the same as `const` since there is no local-scope equivalent of `const`. I am not super adamant about this though, we can be more lenient here or allow these in local scope altogether if you think that's more intuitive.

Reply:
I think the main driver is the kind of variable in question, global vs. local. Almost everything about them is different. So I think all declarations on globals (type or const) should work in a way that makes sense for globals, and all declarations on locals should make sense for them.

Reply:
One reason to allow type declarations of globals inside functions is that sometimes, figuring out the type is difficult from a global scope without introducing a new variable. For a dumb example, `f = x->x+1` is very hard to type annotate from global scope without introducing a second global variable.

Reply:
Ok, I will make this a runtime check.
@oscardssmith That usecase is not supported though, the type declaration always needs to be moved to the toplevel in lowering, otherwise inference can't access that type information.